core/src/main/kotlin/com/strumenta/kolasu/transformation/Transformation.kt

```diff
@@ -17,15 +20,26 @@ annotation class Mapped(val path: String = "")
 /**
  * Factory that, given a tree node, will instantiate the corresponding transformed node.
  */
-class NodeFactory<Source>(
-    val constructor: (Source, ASTTransformer, NodeFactory<Source>) -> Node,
+class NodeFactory<Source, Output : Node>(
```

Comment:
I verified that this additional type parameter does not seem to require an extra declaration in the examples I have developed for the demo (the system is still able to infer the value), so that looks good to me

Reply:
The additional type parameter has been introduced to avoid additional casts in 
```kotlin
finally { targetNode -> ...targetNode as SomeNodeSubClass... }
```
but we could remove it in the future if we see that the change causes issues.